File: 0684-redundant-connection/0684-redundant-connection.js
/**
 * @param {number[][]} edges
 * @return {number[]}
 */
var findRedundantConnection = function (edges) {
    let res = [];
    let parent = new Map();

    function find(node) {
        if (!parent.get(node)) {
            parent.set(node, node);
        }

        if (parent.get(node) !== node) {
            parent.set(node, find(parent.get(node)));
        }

        return parent.get(node);
    }

    function union(u, v) {
        const rootU = find(u);
        const rootV = find(v);

        if (rootU !== rootV) {
            parent.set(rootU, rootV);
        } else {
            res = [u, v];
        }
    }

    edges.forEach(([u, v]) => union(u, v));
    return res;
};

File: 0684-redundant-connection/0684-redundant-connection.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        parent = [i for i in range(n+1)]
        rank = [1] * (n+1)
        redundant = [-1, -1]

        def union(x, y):
            px = find(x)
            py = find(y)

            if px != py:
                if rank[px] >= rank[py]:
                    parent[py] = px
                    rank[px] += 1
                else:
                    parent[px] = py
                    rank[py] += 1
            else:
                # storing redundant
                redundant[0] = x
                redundant[1] = y
                rank[px] += 1

        def find(x):
            if x != parent[x]:
                return find(parent[x])
            return x
        
        for u,v in edges: union(u, v)

        return redundant
